File: fileutlityjava/src/main/java/jgappsandgames/me/fileutlityjava/FileUtility.java
package jgappsandgames.me.fileutlityjava;

import android.content.Context;
import android.os.Environment;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Calendar;

/**
 * FileUtility
 *
 * Handles the loading and the saving of the data
 *
 * Version 1.1.0
 * - Adds External File
 * - Adds Cleaning Cache and data Directory
 */
public class FileUtility {
    // Data ----------------------------------------------------------------------------------------
    private static File data;
    private static File cache;

    public static File external;
    public static String path;

    // Static Methods ------------------------------------------------------------------------------
    public static boolean loadFilePaths(Context context) {
        data = context.getFilesDir();
        cache = context.getCacheDir();

        if (new File(data, ".firstrun").isFile()) return false;

        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(new File(data, ".firstrun")));
            writer.write(String.valueOf(Calendar.getInstance().getTimeInMillis()));
            writer.flush();
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return true;
    }

    public static boolean loadFilePaths(Context context, String externalPath) {
        path = externalPath;
        data = context.getFilesDir();
        cache = context.getCacheDir();

        if (new File(data, ".firstrun").isFile()) return false;

        try {
            BufferedWriter writer = new BufferedWriter(new FileWriter(new File(data, ".firstrun")));
            writer.write(String.valueOf(Calendar.getInstance().getTimeInMillis()));
            writer.flush();
            writer.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        return true;
    }

    // Internal File Methods -----------------------------------------------------------------------
    public static File getFilepath() throws FileUtilityException {
        if (data == null) throw new FileUtilityException("Filepaths are not loaded");
        return data;
    }

    public static File getFile(String filename) throws FileUtilityException {
        if (data == null) throw new FileUtilityException("Filepaths are not loaded");
        return new File(data, filename);
    }

    // Cache Methods -------------------------------------------------------------------------------
    public static File getCachePath() throws FileUtilityException {
        if (cache == null) throw new FileUtilityException("Filepaths are not loaded");
        return cache;
    }

    public static File getCache(String filename) throws FileUtilityException {
        if (cache == null) throw new FileUtilityException("Filepaths are not loaded");
        return new File(cache, filename);
    }

    // External File Methods -----------------------------------------------------------------------
    public static File getExternalPath() throws FileUtilityException {
        if (external == null) {
            if (path != null) {
                external = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS), path);
                if (!external.exists()) external.mkdirs();
            } else {
                throw new FileUtilityException("The external Directory cannot be set up");
            }
        }

        return external;
    }

    public static File getExternal(String filename) throws FileUtilityException {
        if (external == null) {
            if (path != null) {
                external = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS), path);
                if (!external.exists()) external.mkdirs();
            } else {
                throw new FileUtilityException("The external Directory cannot be set up");
            }
        }

        return new File(external, filename);
    }
}

File: fileutlitykotlin/src/main/java/jgappsandgames/me/fileutlitykotlin/FileUtility.kt
package jgappsandgames.me.fileutlitykotlin

import android.content.Context
import android.os.Environment

import java.io.BufferedWriter
import java.io.File
import java.io.FileWriter
import java.io.IOException
import java.util.Calendar

/**
 * FileUtility
 *
 * Handles the loading and the saving of the data
 *
 * Version 1.1.0
 * - Adds External File
 * - Adds Cleaning Cache and data Directory
 */

// Data --------------------------------------------------------------------------------------------
private var data: File? = null
private var cache: File? = null

var external: File? = null
var path: String? = null

// Methods -----------------------------------------------------------------------------------------
fun loadFilePaths(context: Context): Boolean {
    data = context.filesDir
    cache = context.cacheDir

    if (File(data, ".firstrun").isFile) return false

    try {
        val writer = BufferedWriter(FileWriter(File(data, ".firstrun")))
        writer.write(Calendar.getInstance().timeInMillis.toString())
        writer.flush()
        writer.close()
    } catch (e: IOException) {
        e.printStackTrace()
    }

    return true
}

fun loadFilePaths(context: Context, externalPath: String): Boolean {
    path = externalPath
    data = context.filesDir
    cache = context.cacheDir

    if (File(data, ".firstrun").isFile) return false

    try {
        val writer = BufferedWriter(FileWriter(File(data, ".firstrun")))
        writer.write(Calendar.getInstance().timeInMillis.toString())
        writer.flush()
        writer.close()
    } catch (e: IOException) {
        e.printStackTrace()
    }

    return true
}

@Throws(FileUtilityException::class)
fun getFilepath(): File {
    if (data == null) throw FileUtilityException("Filepaths are not loaded")
    return data!!
}

@Throws(FileUtilityException::class)
fun getFile(filename: String? = null): File {
    if (data == null) throw FileUtilityException("Filepaths are not loaded")

    return if (filename != null) File(data!!, filename)
    else data!!
}

@Throws(FileUtilityException::class)
fun getCachePath(): File {
    if (cache == null) throw FileUtilityException("Filepaths are not loaded")
    return cache!!
}

@Throws(FileUtilityException::class)
fun getCache(filename: String? = null): File {
    if (cache == null) throw FileUtilityException("Filepaths are not loaded")

    return if (filename != null) File(cache!!, filename)
    else cache!!
}

@Throws(FileUtilityException::class)
fun getExternalPath(): File {
    if (external == null) {
        if (path != null) {
            external = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS), path)
            if (!external!!.exists()) external!!.mkdirs()
        } else {
            throw FileUtilityException("The external Directory cannot be set up")
        }
    }

    return external!!
}

@Throws(FileUtilityException::class)
fun getExternal(filename: String): File {
    if (external == null) {
        if (path != null) {
            external = File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOCUMENTS), path)
            if (!external!!.exists()) external!!.mkdirs()
        } else {
            throw FileUtilityException("The external Directory cannot be set up")
        }
    }

    return File(external!!, filename)
}